sdk/communication/azure-communication-phonenumbers/src/main/java/com/azure/communication/phonenumbers/PhoneNumbersAsyncClient.java

```diff
@@ -0,0 +1,377 @@
+        updatePhoneNumberWithResponse(String phoneNumber, PhoneNumberUpdateRequest update) {
+        try {
+            Objects.requireNonNull(phoneNumber, "'phoneNumber' cannot be null.");
+            Objects.requireNonNull(update, "'update' cannot be null.");
+            return client.updateWithResponseAsync(phoneNumber, update);
+        } catch (RuntimeException ex) {
+            return monoError(logger, ex);
+        }
+    }
+
+    /**
+     * Starts the search for available phone numbers to purchase.
+     *
+     * @param countryCode The ISO 3166-2 country code.
+     * @param searchRequest {@link PhoneNumberSearchRequest} specifying the search request
+     * until it gets a result from the server
+     * @return A {@link PollerFlux} object with the reservation result
+     */
+    @ServiceMethod(returns = ReturnType.LONG_RUNNING_OPERATION)
+    public PollerFlux<PhoneNumberOperation, PhoneNumberSearchResult> beginSearchAvailablePhoneNumbers(
+        String countryCode, PhoneNumberSearchRequest searchRequest) {
+        return beginSearchAvailablePhoneNumbers(countryCode, searchRequest, null);
+    }
+
+    PollerFlux<PhoneNumberOperation, PhoneNumberSearchResult> beginSearchAvailablePhoneNumbers(
+        String countryCode, PhoneNumberSearchRequest searchRequest, Context context) {
+        try {
+            Objects.requireNonNull(countryCode, "'countryCode' cannot be null.");
+            Objects.requireNonNull(searchRequest, "'searchRequest' cannot be null.");
+
+            return new PollerFlux<>(defaultPollInterval,
+                searchAvailableNumbersInitOperation(countryCode, searchRequest, context),
+                pollOperation(),
+                cancelOperation(),
+                searchAvailableNumbersFetchFinalResultOperation());
+
+        } catch (RuntimeException ex) {
+            return PollerFlux.error(ex);
+        }
+    }
+
+    private Function<PollingContext<PhoneNumberOperation>, Mono<PhoneNumberOperation>>
+        searchAvailableNumbersInitOperation(String countryCode, PhoneNumberSearchRequest searchRequest, Context context) {
+        return (pollingContext) -> {
+            return withContext(contextValue -> {
+                if (context != null) {
+                    contextValue = context;
+                }
+                return client.searchAvailablePhoneNumbersWithResponseAsync(countryCode, searchRequest, contextValue)
+                .flatMap((PhoneNumbersSearchAvailablePhoneNumbersResponse response) -> {
+                    pollingContext.setData("operationId", parseIdFromUrl(response.getDeserializedHeaders().getOperationLocation()));
+                    pollingContext.setData("searchId", parseIdFromUrl(response.getDeserializedHeaders().getLocation()));
+                    return client.getOperationAsync(pollingContext.getData("operationId"));
+                });
+            });
+        };
+    }
+
+    private String parseIdFromUrl(String url) {
```

Comment:
I see this is called from service response. Could this url be null ? Should this be validated so we do not throw NPE.

Reply:
Thanks, I will add this, but we will likely remove this logic as there are plans to return the ids in the headers without having the parse.